internal/daemon/api_logs.go

```diff
@@ -0,0 +1,312 @@
+	// cancelled or there are no more logs (in non-follow mode).
+	requestStarted := time.Now().UTC()
+	for {
+		select {
+		case log := <-logs:
+			if log.Time.IsZero() {
+				// Zero-time log means we've consumed all buffered logs
+				if flushFifo() != nil {
+					return
+				}
+				if follow {
+					// Following, wait for more
+					numLogs = 0 // so we don't use the FIFO from here on
+					continue
+				}
+				// Not following, we're done
+				return
+			}
+
+			// Logs are coming faster than we can send them (probably a slow
+			// client), so stop now.
+			if !follow && log.Time.After(requestStarted) {
+				flushFifo()
+				return
+			}
+
+			if numLogs > 0 {
+				// Push through FIFO so we only output the first "n" across
+				// all services.
+				if len(fifo) == cap(fifo) {
+					// FIFO channel full, discard oldest log entry before
+					// writing new one so it doesn't block.
+					<-fifo
+				}
+				fifo <- log
+				continue
+			}
+
+			// Otherwise encode and output log directly.
+			err := encoder.Encode(jsonLog(log))
+			if err != nil {
+				logger.Noticef("error writing logs: %v", err)
+				return
+			}
+			if follow {
+				flushWriter(w)
+			}
+
+		case err := <-errors:
+			logger.Noticef("%s", err)
+			return
+
+		case <-req.Context().Done():
+			return
+		}
+	}
+}
+
+// streamLogs reads and parses logs from the given iterators, merging the
+// log streams and ordering by timestamp. It sends the parsed logs to the
+// logs channel, and any error to the errors channel. It returns when the
+// done channel is closed.
+func streamLogs(
+	itsByName map[string]servicelog.Iterator,
+	logs chan<- servicelog.Entry,
+	errors chan<- error,
+	done <-chan struct{},
+) {
+	// Make a channel and register it with each of the iterators to be
+	// notified when new data comes in. We don't strictly need this when not
+	// following, but it doesn't hurt either.
+	notification := make(chan bool, 1)
+	for _, it := range itsByName {
+		it.Notify(notification)
+	}
+
+	// Make sorted list of service names we have iterators for.
+	var services []string
+	for name := range itsByName {
+		services = append(services, name)
+	}
+	sort.Strings(services)
+
+	// Create an iterator and log parser for each service.
+	iterators := make([]servicelog.Iterator, len(services))
+	parsers := make([]*servicelog.Parser, len(services))
+	for i, name := range services {
+		iterators[i] = itsByName[name]
+		parsers[i] = servicelog.NewParser(iterators[i], logReaderSize)
+	}
+
+	// Slice of next entries for each service
+	nexts := make([]servicelog.Entry, len(services))
+
+	// Main loop: output earliest log per iteration. Stop when done is closed.
+	for {
+		// Try to fetch next log from each service (parser/iterator combo).
+		for i, parser := range parsers {
+			if !nexts[i].Time.IsZero() {
+				continue
+			}
+			if parser.Next() {
+				nexts[i] = parser.Entry()
+			} else if parser.Err() != nil {
+				errors <- fmt.Errorf("error parsing logs: %w", parser.Err())
```

Comment:
unless errors is buffered and guarantee only one send on that channel, then this needs to be protected in a select with done

Reply:
This is gone now that we're returning the error (and `errorChan` now is now 1-buferred in any case).